nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/status/history/RocksDBComponentStatusRepository.java

```diff
@@ -0,0 +1,703 @@
+                    while(byteArrayInputStream.read(key, 0, 8) == 8) {
+                        byte[] gc = db.get(key);
+                        GarbageCollectionStatus garbageCollectionStatus = unmarshall(gc, StandardGarbageCollectionStatus.class);
+                        if(garbageCollectionStatus == null || garbageCollectionStatus.getTimestamp().before(start) || garbageCollectionStatus.getTimestamp().after(end)) {
+                            continue;
+                        }
+                        history.addGarbageCollectionStatus(garbageCollectionStatus);
+                    }
+                }
+            }
+
+            return history;
+        } catch (RocksDBException | IOException e) {
+            logger.error("Could not get Garbage Collection History", e);
+            return null;
+        }
+    }
+
+    /**
+     * Get the Status history for the requested component
+     *
+     * @param componentId               the component to get the history of
+     * @param includeCounters           whether to include counters
+     * @param defaultMetricDescriptors  descriptors for the metrics
+     * @return the status history
+     */
+    private StatusHistory getStatusHistory(String componentId, boolean includeCounters, Set<MetricDescriptor<?>> defaultMetricDescriptors) {
+        try {
+            byte[] componentKey = componentId.getBytes(UTF8);
+
+            byte[] historyIds = db.get(componentsHandle, componentKey);
+
+            if(historyIds == null || historyIds.length == 0) {
+                return createEmptyStatusHistory();
+            }
+
+            Map<Long, StatusSnapshot> snapshotMap = new HashMap<>();
+            List<Future> queryFutures = new ArrayList<>();
+
+            try(ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(historyIds)) {
+                byte[] read = new byte[8];
+                while(byteArrayInputStream.read(read, 0, 8) == 8) {
+                    byte[] key = new byte[read.length];
+                    System.arraycopy(read, 0, key, 0, key.length);
+                    queryFutures.add(executorService.submit(() -> {
+                        byte[] status;
+                        try {
+                            status = db.get(key);
+                            if(status != null) {
+                                StatusSnapshot statusSnapshot = unmarshall(status, StandardStatusSnapshot.class);
+                                if(statusSnapshot != null) {
+                                    snapshotMap.put(statusSnapshot.getTimestamp().getTime() / 1000L, statusSnapshot);
+                                }
+                            }
+                        } catch(RocksDBException e) {
+                            logger.warn("Could not retrieve status snapshot", e);
+                        }
+                    }));
+                }
+
+                // wait for futures to complete... throws TimeoutException
```

Comment:
It's probably a little bit nicer to use CompletableFutures here so you can easily wait or get a callback when all of the futures are done:

```
CompletableFuture[] futures = ...
CompletableFuture.allOf(futures)
   .thenApply(//the rest);
```
or
```
CompletableFuture.allOf(futures).join();
```

source: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#allOf-java.util.concurrent.CompletableFuture...-